edx_event_bus_kafka/internal/consumer.py

```diff
@@ -129,10 +130,30 @@ def _shut_down(self):
         """
         self._shut_down_loop = True
 
-    def consume_indefinitely(self):
+    def consume_indefinitely(self, offset_timestamp=None):
         """
         Consume events from a topic in an infinite loop.
+
+        Arguments:
+            offset_timestamp(datetime): reset the offsets of the consumer partitions to this timestamp before consuming.
         """
+
+        def reset_offsets(consumer, partitions):
+            # This is a callback method used on consumer assignment to handle offset reset logic.
+            # We do not want to attempt to change offsets if the offset is None.
+            if offset_timestamp is None:
+                return
+
+            # Get the offset from the epoch in seconds.
+            offset_timestamp_s = int(offset_timestamp.timestamp())
+            # We construct partitions with the epoch timestamp in the offset position.
+            timestamp_partitions = [TopicPartition(self.topic, p.partition, offset_timestamp_s) for p in partitions]
+
+            partitions_with_offsets = consumer.offsets_for_times(timestamp_partitions, timeout=1.0)
```

Comment:
I believe we'll also need to check the `error` field on the partitions: https://github.com/confluentinc/confluent-kafka-python/pull/268/files#diff-34dc92cd663b144203a9252b1eeb69bc3f10cdaaea64f9ec379503be3cc62485R1001

> returns: list of topic+partition with offset field set and possibly error set

Reply:
Not to kick the can too far down the road, but I think this might also be handled by Kafka itself when it looks through the list of partitions?

Reply:
Hmm... that would be good, but I can't find anything reassuring in confluent-kafka or librdkafka in that direction. Here's librdkafka's implementation of assign: https://github.com/edenhill/librdkafka/blob/v1.9.2/src/rdkafka_subscription.c#L119

It's possible we'd get an error during poll or something else, but I'd rather stay on the safe side and check that all the error attributes are None. Otherwise we might end up with something goofy like _actually_ having timestamps in the offsets, etc.

Reply:
Okay, will add some error handling around this, then.